### DISCUSSION TITLE
T95w2 box (S905W2 Experimental)

### DISCUSSION THREAD

**Discussion by sadptech** [devmfc/debian-on-amlogic#17] [Q&A]
Hello guys,

Wanted to ask why I get kernel panic no rootfs found then i boot from USB or microsd. Tried multiple SD cards and USB flash drives same kernel panic...

Box is s905w2 box is there a way to detect SD on boot and mount it automatically or manually?

![IMG_20231217_160133](https://github.com/devmfc/debian-on-amlogic/assets/64962997/c66c58ac-137c-4b9f-b90a-3dac739a1bd6)

**Comment by devmfc** [comment]
It can't find the rootfs partition. Looks like you have two usb flash drives in your box while booting? Remove all USB drives from the box before booting. Otherwise try to use root=/dev/sdb2 in your boot.config.

**Comment by sadptech** [comment]
ok for some reason it did not like usb flash and sd cards i tried sandisk and kingston....

with generic sd china it boot fast :D

sorry.. i did not think that sd card brand means something here omg...

anyway, image boots fine now, but it has no DE :( tried to install xfce sadly does work, im really newbie here..

if there is a way you can add xfce builds in future would be nice, or atleast some script to post install it :)

edit:
if there is an easy way i could extract dtb file from android somehow? and you could add this t95w box to supported list leds and sensors seem not work now :)


![image](https://github.com/devmfc/debian-on-amlogic/assets/64962997/64ae65e6-0220-4971-9774-2fa233110b69)
![image](https://github.com/devmfc/debian-on-amlogic/assets/64962997/0da02f4b-43b8-458c-90b0-8163bcb7266d)

**Reply by devmfc** [reply]
Great to see. Sensors have, in this case, nothing to do with DTB, drivers must be modified for S4 sensors. 

No, I will never release desktop images. It is not necessary also. My images are just standard minimal Ubuntu/Debian images, so you can install whatever is supported by Debian/Ubuntu. So [this link](https://www.google.com/search?q=ubuntu+how+to+install+xfce+desktop) will give you anything you need.

IMHO, these boxes are waaay to slow for desktop usage. And there is no real support for HDMI for s905w2 SOC's anyway.

**Reply by ni-h** [reply]
Hi, I had a very similar experience, just wanted to share it. 

My device is a T95W2 2/16GB with an S905W2 (AliExpress: /item/1005005876016495.html). Attempting to boot directly off a uSD (SanDisk Ultra SDHC 32GB) did not work and gave the same Kernel panic as shown in the original post. Using an old, cheap, generic uSD-USB 2.0 adapter, it booted just fine. I read in some other forums that somehow, the code does not expect the fast speed that the uSD and the reader provide, so using slower, generic cards or adapters works better (i dont remember what or where exactly it was though so consider it hearsay).

What didnt work for me is WiFi. Using ifconfig, wlan0 does not show up and lshw only shows the ethernet. Checking the board, I can see the wireless module is an sp6330 module. If there is anything you want me to try regarding this, or logs, let me know.

Secondly, I wanted to confirm that the experimental install to eMMC script worked flawlessly, too. No issues here.

Thirdly, the install minimal desktop script did not work, it installed everything but I was not able to start the desktop environment. I would have liked to try just how fast (slow) it is :)

And lastly: Using the multiboot script worked just as well. So for example, I can have CoreELEC on the uSD and it will boot straight to that if the uSD is inserted.

Thank you for sharing your work with us.

**Reply by sadptech** [reply]
Yes just wanted to add that it seems like bt/wifi not work. With these cheap boxes its mixed bag which chips we get... Xfce not work or i dont know how.. but for headless LAN system it is nice.

**Reply by devmfc** [reply]
I would expect the sp6330 to work... You could try to install the full firmware package via `apt-get install firmware-linux`. If that doesn't work you could try to see if `dmesg` complains about something... Otherwise report the output of the following command: `( vendev=( $( cat /sys/bus/mmc/devices/mmc2:*/{vendor,device} ) ) && echo "wifi SDIO ID: ${vendev[0]}:${vendev[1]}" )`

There is no proper HDMI support for S904X/S905W2 in mainline, so desktop usage will not work.

Thanks for confirming multiboot does indeed work with CoreELEC, good to hear.

**Reply by ni-h** [reply]
Thanks for your reply. I installed the full firmware package but it did not change anything. 

the command you suggested yielded no results: 
`tvbox:~#( vendev=( $( cat /sys/bus/mmc/devices/mmc2:*/{vendor,device} ) ) && echo "wifi SDIO ID: ${vendev[0]}:${vendev[1]}" )
cat: '/sys/bus/mmc/devices/mmc2:*/vendor': No such file or directory
cat: '/sys/bus/mmc/devices/mmc2:*/device': No such file or directory`


The full dmesg can be found [here](https://pastebin.com/kTWswkU5).
I thought there was nothing in the dmesg, but with respect to your suggested command, this line stands out(?): 
`[    0.326559] mmc2: Failed to initialize a non-removable card`

Is this where the wifi card is supposed to be initialised? Could the root of the cause be in the dtb?

**Reply by Adney** [reply]
> I would expect the sp6330 to work... You could try to install the full firmware package via `apt-get install firmware-linux`. If that doesn't work you could try to see if `dmesg` complains about something... Otherwise report the output of the following command: `( vendev=( $( cat /sys/bus/mmc/devices/mmc2:*/{vendor,device} ) ) && echo "wifi SDIO ID: ${vendev[0]}:${vendev[1]}" )`

wifi SDIO ID: 0x02d0:0x4330

**Reply by devmfc** [reply]
So very strange... Your box returns the SDIO ID, the other one doesn't... same box or not...?
But it can only retun an SDIO Id if the SDIO card is powered on. Can you post the dmesg after a fresh reboot? Because I have other boxes with a broadcom 4330 wifi module and they work fine....

**Reply by Adney** [reply]
This worked on the v6.1.57 kernel and did not work on later ones (v.6.6.1 -did not work...)

Either it turns out to detect emmc and write the OS to it (this is in versions starting from v6.6.1 ). Or it turns out to detect wifi (in versions 6.1.57 - v6.1.25), but I couldn't get it to work.

This box has a problem with the card reader, or the SD card settings in dtb. Almost no SD card allows you to boot. I have only one card that everything works with.

**Comment by dmotkan** [comment]
The problem with the card reader is most likely due to incorrect layout of the printed circuit board and the absence of power decoupling capacitors. Lots of interference. Therefore, only low-speed SD cards work.

HDMI output works on the latest builds

**Comment by Adney** [comment]
I tried the latest release 6.6.23 with the T95W2 Y4 configuration. The LED screen is working, but the wi-fi/bt chip is not detected. Then, based on this dtb (where the LED screen works), and another dtb where the wi-fi chip is determined, I assembled my dtb.
[1meson-s4-t95w2.zip](https://github.com/devmfc/debian-on-amlogic/files/14892738/1meson-s4-t95w2.zip)

Now I have a working LED display, and if I enter the command
`( vendev=( $( cat /sys/bus/mmc/devices/mmc2:*/{vendor,device} ) ) && echo "wifi SDIO ID: ${vendev[0]}:${vendev[1]}" )`
That is determined by the wi-Fi/bt chip:
**wifi SDIO ID: 0x02d0:0x4330**

![8888](https://github.com/devmfc/debian-on-amlogic/assets/18748732/eb4e1399-e99d-4ae8-accf-096eb6b85736)
But the wi-fi itself is still not working.

**dmesg | grep brcm**
[ 6.044672] brcmfmac: _brcmf_fw_alloc_request: using brcm/brcmfmac4330-sdio for chip BCM4330/4_
[ 6.050322] usbcore: _registered new interface driver brcmfmac_
[ 6.123218] brcmfmac mmc2:0001:1: _Direct firmware load for brcm/brcmfmac4330-sdio.clm_blob failed with error -2_

tvbox:~#vendev
**wifi SDIO ID: 0x02d0:0x4330**

tvbox:~#./wifi-connect.sh
Unknown device "/sys/class/net/wlan0": No such device
Connect to wifi network wlan0:
no wlan interface found

**I need help!**
